csharp/src/Ice/TlsClientOptions.cs

```diff
@@ -0,0 +1,97 @@
+        /// </summary>
+        public LocalCertificateSelectionCallback? ClientCertificateSelectionCallback { get; set; }
+
+        /// <summary>Gets or set the list of SSL protocols to enable for outgoing connections, the default is None,
+        /// that allows the OS to choose the best protocol to use. It can be set to Tls12, Tls13, None or a combination
+        /// of those, other values are not accepted.</summary>
+        public SslProtocols? EnabledSslProtocols { get; set; }
+
+        /// <summary>Gets or sets the certificates collection that will be used as trusted certificate authorities
+        /// to verify the server certificate used for authentication. Setting this is incompatible with setting
+        /// <see cref="ServerCertificateValidationCallback"/>.</summary>
+        public X509Certificate2Collection? ServerCertificateCertificateAuthorities
+        {
+            get => _serverCertificateCertificateAuthorities;
+            set
+            {
+                if (ServerCertificateValidationCallback != null)
+                {
+                    throw new ArgumentException(
+                        @$"Using the {nameof(ServerCertificateCertificateAuthorities)} is incompatible with using the {
+                           nameof(ServerCertificateValidationCallback)}");
+                }
+                _serverCertificateCertificateAuthorities = value;
+            }
+        }
+
+        /// <summary>Gets or sets the callback that will be used to verify the server certificate. Setting this is
+        /// incompatible with setting <see cref="ServerCertificateCertificateAuthorities"/> and IceSSL.CAs
+        /// configuration property.</summary>
+        public RemoteCertificateValidationCallback? ServerCertificateValidationCallback
+        {
+            get => _serverCertificateValidationCallback;
+            set
+            {
+                if (CertificateRevocationCheckMode != null)
+                {
+                    throw new ArgumentException(
+                        @$"Using the {nameof(ServerCertificateValidationCallback)} is incompatible with using the {
+                           nameof(CertificateRevocationCheckMode)}");
+                }
+
+                if (ServerCertificateCertificateAuthorities != null)
+                {
+                    throw new ArgumentException(
+                        @$"Using the {nameof(ServerCertificateValidationCallback)} is incompatible with using the {
+                           nameof(ServerCertificateCertificateAuthorities)}");
+                }
+
+                _serverCertificateValidationCallback = value;
+            }
+        }
+
+        private X509RevocationMode? _certificateRevocationCheckMode;
+        private X509Certificate2Collection? _serverCertificateCertificateAuthorities;
+        private RemoteCertificateValidationCallback? _serverCertificateValidationCallback;
+    }
+}
```

Comment:
Are all these checks necessary. Also they don't seem exhaustive. If a user sets `ServerCertificateValidationCallback` after settings `CertificateRevocationCheckMode` it breaks. If we need a to check, we should maybe call a validate() function that checks all combinations?

Reply:
The the reasoning is that if you first set CertificateRevocationCheckMode, and then a custom verification callback, the revocation mode will be ignored as the validation is up to the custom callback. So the validation here depends on the order of the settings, that being set I have remove the revocation mode as suggested by Benot

Reply:
What if I set them all in the constructor. How is the ordering handled, declaration order? I would assume you would typically want to just set everything in one shot?

Reply:
You cannot, there is only a default constructor that doesn't set any properties, you must use the c# object initializer syntax and set whatever properties you want in the order you want, this avoid the need for keyword arguments in the constructor.

So far I just remove the revocation mode property and the remaining validation is just to check that you don't set both CAs and validation callback, the other combinations are fine, so I will just keep this check in the property setter.

Reply:
Ideally when we move to C# 9 this will be record types with init only properties